dwa_local_planner/src/dwa_planner.cpp

```diff
@@ -262,37 +348,40 @@ namespace dwa_local_planner {
       global_plan_[i] = new_plan[i];
     }
 
+    std::vector<geometry_msgs::PoseStamped> front_global_plan = global_plan_;
+    const auto result = cropPlanToReachableGoal(global_pose, footprint_spec, front_global_plan);
+    if (result != mbf_msgs::ExePathResult::SUCCESS) {
+      return result;
+    }
+
+    const geometry_msgs::PoseStamped goal_pose = front_global_plan.back();
+    const double dist_to_goal = mbf_utility::distance(global_pose, goal_pose);
+
     obstacle_costs_.setFootprint(footprint_spec);
 
     path_align_costs_.setTargetPoses(global_plan_, global_pose);
 
     // costs for going away from path
-    path_costs_.setTargetPoses(global_plan_);
-    alignment_costs_.setTargetPoses(global_plan_);
+    path_costs_.setTargetPoses(front_global_plan);
+    alignment_costs_.setTargetPoses(front_global_plan);
 
     // costs for not going towards the local goal as much as possible
-    goal_costs_.setTargetPoses(global_plan_);
+    goal_costs_.setTargetPoses(front_global_plan);
 
     // alignment costs
-    geometry_msgs::PoseStamped goal_pose = global_plan_.back();
-
-    Eigen::Vector3f pos(global_pose.pose.position.x, global_pose.pose.position.y, tf2::getYaw(global_pose.pose.orientation));
-    double sq_dist =
-        (pos[0] - goal_pose.pose.position.x) * (pos[0] - goal_pose.pose.position.x) +
-        (pos[1] - goal_pose.pose.position.y) * (pos[1] - goal_pose.pose.position.y);
 
     // we want the robot nose to be drawn to its final position
     // (before robot turns towards goal orientation), not the end of the
     // path for the robot center. Choosing the final position after
     // turning towards goal orientation causes instability when the
     // robot needs to make a 180 degree turn at the end
-    std::vector<geometry_msgs::PoseStamped> front_global_plan = global_plan_;
-    double angle_to_goal = atan2(goal_pose.pose.position.y - pos[1], goal_pose.pose.position.x - pos[0]);
+    double angle_to_goal = atan2(goal_pose.pose.position.y - global_pose.pose.position.y, goal_pose.pose.position.x - global_pose.pose.position.x);
 
     double forward_point_distance = forward_point_distance_;
     const auto cos_angle_to_goal = cos(angle_to_goal);
     const auto sin_angle_to_goal = sin(angle_to_goal);
-    if (sq_dist < MIN_GOAL_DIST_SQ) {
+    const auto sq_dist = dist_to_goal * dist_to_goal;
+    if (sq_dist * sq_dist < MIN_GOAL_DIST_SQ) {
```

Comment:
```suggestion
    if (sq_dist < MIN_GOAL_DIST_SQ) {
```
F: `dist_to_goal^4`?